### ISSUE TITLE
Optimize results collection from Sauce Labs mediated browsers

### ISSUE THREAD

**Issue by jugglinmike** [web-platform-tests/results-collection#546]
Earlier this year, we identified a bottleneck in the communication channel with Sauce Labs. To date, we have accepted delays in the time-to-results for browsers tested via this service, but there is a way to improve this.

For context: the WPT CLI creates a dedicated "tunnel" with every invocation, and this works well in the context it was designed for (i.e. pull request validation, or the intermittent evaluation of a fraction of WPT).

However, due to API rate limits enforced by Sauce Labs, this method is not appropriate for the purposes of parallelized results collection. It limits the number of tests we can run in parallel to about 4. Results collection through this service is fundamentally slow due to the introduction of network delays, and this restriction lengthens the time required to collect results even further. As of this writing, we only collect results from Edge and Safari once every two days (compare that with once every six hours for locally-run browsers).

We could decrease time-to-results significantly by increasing parallelism, but to do that, we will need to extend the testing infrastructure with a central service that maintains a single shared tunnel to Sauce Labs. This would allow many machines to run in parallel without expanding our usage of the Sauce Labs API beyond the limit.

- Provisioning an additional server and configuring it to maintain a persistent connection with Sauce Labs (using the "Sauce Connect" binary)
- Extending the WPT CLI to allow users to specify the URL to which WebDriver traffic should be sent
- Extending the "worker" configuration to run the WPT server with a unique hostname and to register this name with the proxy (e.g. via DNS or via hostfile)

Given the project's larger goals to run the very latest browser/WebDriver combinations, it is not clear for how much longer we will rely on the Sauce Labs service. For that reason, we have de-prioritized this effort. I'm filing this issue to make our priorities clear and to offer initial thoughts on implementation when/if we decide to take this on.

**Comment by jugglinmike** [comment]
@jgraham One of the tasks listed above is a modification of the WPT CLI:

> - Extending the WPT CLI to allow users to specify the URL to which WebDriver traffic should be sent

[That value is currently hard-coded](https://github.com/w3c/web-platform-tests/blob/e6756de059f68d445306ac35a6b8e80633d75925/tools/wptrunner/wptrunner/browsers/sauce.py#L77), so I'd like to implement a new command-line argument for specifying another location.  I can't say when/if we will do this work, but in advance of that, it would be helpful to know if you would support such an extension.

**Comment by jgraham** [comment]
Yes.

**Comment by jugglinmike** [comment]
Great; thanks!

**Comment by jugglinmike** [comment]
I experimented with the HTTP API requests issued by the Sauce Connect binary and found that the rate limit is enforced on a per-account basis. This may sound fairly obvious, but Sauce Labs allows users to create "sub-accounts", and my research proved that each such account has a dedicated rate limitation. I assigned a dedicated sub-account to each of 20 workers, and the subsequent results collection completed without error.

In other words: we were able to extend parallelization on Sauce Labs by a factor of 5 without the complex extension described above. That obviates the need for such an extension, so I will close this issue.
